### PR TITLE
[UnitaryHack2025] Implement all-to-all and simple Dicke State generation circuits

### LINKED ISSUE #1655: Improved circuit depth for Dicke states in all-to-all connectivity
During the unitaryHACK 2025, we decided to split issue #1620 into two and reward one solution for each issue.

### PR DESCRIPTION
**Pull request by joan-ainaud**
Closes #1655 
Implemented Dicke State generation with all-to-all connectivity option. I see @AlejandroSopena implemented the first part already, should I adapt the second part to extend his?

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

### CODE REVIEW

**Comment by renatomello** [comment]
@joan-ainaud Can you comment on #1655 so I can assign you to it? Thanks.

**Comment by renatomello** [comment]
@joan-ainaud Could you adapt your solution to the same pattern as the recently merged solution in #1650 ?

File: src/qibo/models/encodings.py [hunk 1]
```diff
@@ -1485,69 +1485,137 @@ def _get_phase_gate_correction_sparse(
...
-
     Returns:
         :class:`qibo.models.circuit.Circuit` : Circuit that prepares :math:`\\ket{D_{k}^{n}}`.
-
     References:
         1. Andreas Bärtschi and Stephan Eidenbenz, *Deterministic preparation of Dicke states*,
         `22nd International Symposium on Fundamentals of Computation Theory, FCT'19, 126-139  (2019)
         <https://doi.org/10.1007/978-3-030-25027-0_9>`_.
+        2. Andreas Bärtschi and Stephan Eidenbenz, *Short-Depth Circuits for Dicke State Preparation*,
+        `IEEE International Conference on Quantum Computing & Engineering (QCE), 2022, pp. 87--96
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
        `IEEE International Conference on Quantum Computing & Engineering (QCE), 87--96 (2022)
```

File: src/qibo/models/encodings.py [hunk 2]
```diff
@@ -1485,69 +1485,137 @@ def _get_phase_gate_correction_sparse(
...
+                }
+            )
+        disjoint_sets = list(
+            reversed(disjoint_sets)
+        )  # we reverse to have in ascending order of tier
+
+        trees = disjoint_sets.copy()
+        # combine lowest tier trees into one tree
+        print(": ")
+        print(list(id(t) for t in trees))
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
```

File: src/qibo/models/encodings.py [hunk 3]
```diff
@@ -1485,69 +1485,137 @@ def _get_phase_gate_correction_sparse(
...
+        while len(trees) > 1:
+            first_smallest = trees.pop(0)
+            second_smallest = trees.pop(0)
+            second_smallest["tier"] += first_smallest["tier"]
+            second_smallest["children"].append(first_smallest)
+            new = second_smallest
+            # put new combined tree in list mantaining ordering
+            trees.insert(sum(x["tier"] < new["tier"] for x in trees), new)
+            print(list(id(t) for t in trees))
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
```

File: src/qibo/models/encodings.py [hunk 4]
```diff
@@ -1485,69 +1485,135 @@ def _get_phase_gate_correction_sparse(
...
-def dicke_state(nqubits: int, weight: int, **kwargs):
+def dicke_state(nqubits: int, weight: int, all_to_all: bool = False, **kwargs):
     """Create a circuit that prepares the Dicke state :math:`\\ket{D_{k}^{n}}`.
 
     The Dicke state :math:`\\ket{D_{k}^{n}}` is the equal superposition of all :math:`n`-qubit
     computational basis states with fixed-Hamming-weight :math:`k`.
-    The circuit prepares the state deterministically with :math:`O(k \\, n)` gates and :math:`O(n)` depth.
-
+    The circuit prepares the state deterministically with :math:`O(k \\, n)` gates and :math:`O(n)` depth,
+    or :math:`O(k \\log\\frac{n}{k})` for all-to-all connectivity.
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
    or :math:`O(k \\log\\frac{n}{k})` depth under the assumption of ``all-to-all`` connectivity.

```

File: src/qibo/models/encodings.py [hunk 5]
```diff
@@ -1485,69 +1485,136 @@ def _get_phase_gate_correction_sparse(
...
+    The circuit prepares the state deterministically with :math:`O(k \\, n)` gates and :math:`O(n)` depth,
+    or :math:`O(k \\log\\frac{n}{k})` depth under the assumption of ``all-to-all`` connectivity.
 
     Args:
         nqubits (int): number of qubits :math:`n`.
         weight (int): Hamming weight :math:`k` of the Dicke state.
+        all_to_all (bool): Whether to use the shorter depth implementation from reference
         kwargs: Additional arguments for Circuit initialization.
-
     Returns:
```
Comment (renatomello) [inline src/qibo/models/encodings.py:1501]:
```suggestion

    Returns:
```

File: src/qibo/models/encodings.py [hunk 6]
```diff
@@ -1485,69 +1485,136 @@ def _get_phase_gate_correction_sparse(
...
         :class:`qibo.models.circuit.Circuit` : Circuit that prepares :math:`\\ket{D_{k}^{n}}`.
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion

    Returns:
        :class:`qibo.models.circuit.Circuit` : Circuit that prepares :math:`\\ket{D_{k}^{n}}`.
```

File: src/qibo/models/encodings.py [hunk 7]
```diff
@@ -1485,69 +1485,137 @@ def _get_phase_gate_correction_sparse(
...
     Args:
         nqubits (int): number of qubits :math:`n`.
         weight (int): Hamming weight :math:`k` of the Dicke state.
+        all_to_all (bool): Whether to use the shorter depth implementation from reference
         kwargs: Additional arguments for Circuit initialization.
 
     Returns:
         :class:`qibo.models.circuit.Circuit` : Circuit that prepares :math:`\\ket{D_{k}^{n}}`.
-
     References:
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion

    References:
```

File: src/qibo/models/encodings.py [hunk 8]
```diff
@@ -1485,16 +1485,18 @@ def _get_phase_gate_correction_sparse(
...
     The Dicke state :math:`\\ket{D_{k}^{n}}` is the equal superposition of all :math:`n`-qubit
     computational basis states with fixed-Hamming-weight :math:`k`.
-    The circuit prepares the state deterministically with :math:`O(k \\, n)` gates and :math:`O(n)` depth.
+    The circuit prepares the state deterministically with :math:`O(k \\, n)` gates and :math:`O(n)` depth,
+    or :math:`O(k \\log\\frac{n}{k})` depth under the assumption of ``all-to-all`` connectivity.
 
     Args:
         nqubits (int): number of qubits :math:`n`.
         weight (int): Hamming weight :math:`k` of the Dicke state.
+        all_to_all (bool): Whether to use the shorter depth implementation from reference
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
        all_to_all (bool, optional): If ``False``, uses implementation from Ref. [1].
            If ``True``, uses shorter-depth implementation from Ref. [2].
            Defaults to ``False``.
```

File: src/qibo/models/encodings.py [hunk 9]
```diff
@@ -1504,50 +1506,117 @@ def dicke_state(nqubits: int, weight: int, **kwargs):
         1. Andreas Bärtschi and Stephan Eidenbenz, *Deterministic preparation of Dicke states*,
         `22nd International Symposium on Fundamentals of Computation Theory, FCT'19, 126-139  (2019)
         <https://doi.org/10.1007/978-3-030-25027-0_9>`_.
+        2. Andreas Bärtschi and Stephan Eidenbenz, *Short-Depth Circuits for Dicke State Preparation*,
+        `IEEE International Conference on Quantum Computing & Engineering (QCE), 87--96 (2022)
+        <https://doi.org/10.1109/QCE53715.2022.00027>`_.
     """
     if weight < 0 or weight > nqubits:
         raise ValueError(f"weight must be between 0 and {nqubits}, got {weight}")
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
        raise_error(ValueError, f"weight must be between 0 and {nqubits}, but got {weight}.")
```

File: src/qibo/models/encodings.py [hunk 10]
```diff
@@ -1504,50 +1506,117 @@ def dicke_state(nqubits: int, weight: int, **kwargs):
...
+        # Start with |0⟩^(n-k) |1⟩^k
+        circuit.add(gates.X(qubit) for qubit in range(nqubits - weight, nqubits))
 
-    # Recursively build the unitary U_n,k
-    for m in range(weight, 0, -1):
-        # Add SCS_{m,m-1} acting on last m qubits
-        _add_scs_gate(circuit, m, m - 1)
+        _add_dicke_unitary_gate(circuit, range(nqubits), weight)
+
+    else:
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
        _add_dicke_unitary_gate(circuit, range(nqubits), weight)
        
        return circuit

```

Comment (renatomello) [reply src/qibo/models/encodings.py]: `else` can be made implicitly. You should then remove one indent level from all the code below with exception of the last return

File: src/qibo/models/encodings.py [hunk 11]
```diff
@@ -1504,50 +1506,117 @@ def dicke_state(nqubits: int, weight: int, **kwargs):
...
+            disjoint_sets.append(
+                {
+                    "qubits": list(range(nqubits - nmodk, nqubits)),
+                    "tier": nmodk,
+                    "children": [],
+                }
+            )
+        disjoint_sets = list(
+            reversed(disjoint_sets)
+        )  # we reverse to have in ascending order of tier
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
        # reverse to have in ascending order of tier
        disjoint_sets = list(reversed(disjoint_sets))
```

File: src/qibo/models/encodings.py [hunk 12]
```diff
@@ -1504,50 +1506,117 @@ def dicke_state(nqubits: int, weight: int, **kwargs):
...
+            first_smallest = trees.pop(0)
+            second_smallest = trees.pop(0)
+            second_smallest["tier"] += first_smallest["tier"]
+            second_smallest["children"].append(first_smallest)
+            new = second_smallest
+            # put new combined tree in list mantaining ordering
+            trees.insert(sum(x["tier"] < new["tier"] for x in trees), new)
+
+        root = trees[0]
+        # We initialize |0>^|1>^k  bitstring at root qubits
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
        # We initialize |0>^(n-k)|1>^k  bitstring at root qubits
```

File: src/qibo/models/encodings.py [hunk 13]
```diff
@@ -1504,50 +1506,117 @@ def dicke_state(nqubits: int, weight: int, **kwargs):
...
+        for node in disjoint_sets:
+            _add_dicke_unitary_gate(
+                circuit, node["qubits"], min(weight, len(node["qubits"]))
+            )
 
     return circuit
 
 
-def _add_scs_gate(circuit: Circuit, n: int, k: int):
+def _add_scs_gate(circuit: Circuit, qubits: List[int], k: int):
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
def _add_scs_gate(circuit: Circuit, qubits: List[int], weight: int):
```
and inside the function where applicable

File: src/qibo/models/encodings.py [hunk 14]
```diff
@@ -1504,50 +1506,117 @@ def dicke_state(nqubits: int, weight: int, **kwargs):
...
-    Acts on the last k+1 qubits of the circuit.
+    Implements the SCS_{n,k} unitary from Definition 3 of the paper [1].
+    Acts on the last k+1 qubits of the n passed qubits.
     """
     if k == 0:
         return  # SCS_{n,0} is identity
 
-    last_qubit = n - 1
-    # first_qubit = last_qubit - k
+    n = len(qubits)
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
    nqubits = len(qubits)
```

Avoiding non-descriptive variable names

File: src/qibo/models/encodings.py [hunk 15]
```diff
@@ -1557,6 +1626,85 @@ def _add_scs_gate(circuit: Circuit, n: int, k: int):
         circuit.add(gates.CNOT(target_qubit, last_qubit))
 
 
+def _add_dicke_unitary_gate(circuit: Circuit, qubits: List[int], k: int):
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
def _add_dicke_unitary_gate(circuit: Circuit, qubits: List[int], weight: int):
```

Same here

File: src/qibo/models/encodings.py [hunk 16]
```diff
@@ -1557,6 +1626,85 @@ def _add_scs_gate(circuit: Circuit, n: int, k: int):
...
+    """In-place addition to ``circuit`` of a U_{n,k} gate from Definition 2 of the paper [1]."""
+    n = len(qubits)
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
    nqubits = len(qubits)
```

Same here

File: src/qibo/models/encodings.py [hunk 17]
```diff
@@ -1557,6 +1626,85 @@ def _add_scs_gate(circuit: Circuit, n: int, k: int):
...
+        _add_scs_gate(circuit, qubits[:m], m - 1)
+
+
+def _add_wbd_gate(
+    circuit: Circuit,
+    first_register: List[int],
+    second_register: List[int],
+    n: int,
+    m: int,
+    k: int,
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
    nqubits: int,
    m: int,
    weight: int,
```

Same here (including  `m`)

File: src/qibo/models/encodings.py [hunk 18]
```diff
@@ -1557,6 +1626,85 @@ def _add_scs_gate(circuit: Circuit, n: int, k: int):
...
+):
+    """In-place addition of a Weight Distribution Block (WBD) to ``circuit``.
+    Implements the :math:`WBD^{n,m}_k` unitary from Definition 2 of the paper [2].
+    Only acts on first_register and second_register.
+    Our circuit is mirrored, as paper [2] uses a top-bottom circuit <-> right-left bitstring convention
+    """
+
+    if m > n / 2:
+        raise (ValueError("m must not be greater than n-m"))
```
Comment (renatomello) [inline src/qibo/models/encodings.py]:
```suggestion
    if m > n / 2:
        raise_error(ValueError, "``m`` must not be greater than ``n - m``."))
```

File: src/qibo/models/encodings.py [hunk 19]
```diff
@@ -1530,95 +1530,95 @@ def dicke_state(nqubits: int, weight: int, all_to_all: bool = False, **kwargs):
...
+        for it in range(nqubits // weight)
+    ]
+    nmodk = nqubits % weight
+    if nmodk != 0:
+        disjoint_sets.append(
             {
-                "qubits": list(range(weight * it, weight * (it + 1))),
-                "tier": weight,
+                "qubits": list(range(nqubits - nmodk, nqubits)),
+                "tier": nmodk,
```
Comment (renatomello) [inline src/qibo/models/encodings.py:1547]:
```suggestion
                "qubits": list(range(nqubits - n_mod_k, nqubits)),
                "tier": n_mod_k,
```

**Comment by renatomello** [comment]
@joan-ainaud if you could fix the coverage issue, then we just wait for @AlejandroSopena's review

**Comment by joan-ainaud** [comment]
Yes! Just to make sure it's ok, as I'm a bit new to this PR workflow, but I ignored the check in _add_wbd_gate from coverage (with "# pragma: no cover", right?), as it can't trigger within dicke_state()

**Comment by renatomello** [comment]
> Yes! Just to make sure it's ok, as I'm a bit new to this PR workflow, but I ignored the check in _add_wbd_gate from coverage (with "# pragma: no cover", right?), as it can't trigger within dicke_state()

`pragma: no cover` is used as a last-resort solution when some lines cannot be tested for one reason or another. This is not the case here. The first line not covered is the one that returns an empty circuit when $k = 0$. Thus, you could add a test in the test function that checks if the `dicke_state` function is returning an empty circuit when $k = 0$. The other missing line is one that checks if $k > n/2$ and raises an error if True. You could also add a test for that (check the existing tests using `pytest.raises`).

**Review by renatomello — APPROVED** [review]

File: tests/test_models_encodings.py [hunk 20]
```diff
@@ -503,6 +504,55 @@ def test_dicke_state(backend, nqubits, weight, all_to_all, density_matrix):
...
+
+            # initial state |0>^n-l |1>^l
+            initial[2**l-1] = 1
+            
+            comb_n_l = math.comb(nqubits, l)
+            for i in range(min(l,mqubits)+1):
+                index = 2**(nqubits-mqubits) * (2**(i) - 1) + 2**(l-i)-1
+                target[index] = np.sqrt(math.comb(mqubits,i)*math.comb(nqubits-mqubits, l-i) / comb_n_l)
+
+            initial = backend.cast(initial, dtype=target.dtype)
```
Comment (renatomello) [inline tests/test_models_encodings.py]:
```suggestion
            initial = backend.cast(initial, dtype=initial.dtype)
```

**Review by AlejandroSopena — APPROVED** [review]
